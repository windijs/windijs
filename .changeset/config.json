{
  "$schema": "https://unpkg.com/@changesets/config@2.1.1/schema.json",
  "changelog": "@changesets/cli/changelog",
  "commit": false,
  "fixed": [["@windijs/core", "windijs"]],
  "linked": [],
  "access": "public",
  "baseBranch": "main",
  "updateInternalDependencies": "patch",
  "ignore": []
}
