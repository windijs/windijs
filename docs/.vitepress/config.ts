import { defineConfig } from "vitepress";
import { mdRenderFilename } from "./markdown";

export default defineConfig({
  lang: "en-US",
  title: "Windi JS",
  cleanUrls: "without-subfolders",
  description: "The Utility-first CSS-in-JS Framework Powered by ES6 Proxy API and TypeScript.",
  markdown: {
    config(md) {
      md.use(mdRenderFilename);
    },
  },
  head: [
    ['link', { rel: "shortcut icon", type: "image/x-icon", href: "/windijs/favicon.ico" }]
  ],
  base: "/windijs/",
  themeConfig: {
    logo: "/logo.png",
    nav: [
      { text: "Guide", link: "/guide/what-is-windijs" },
      { text: "API", link: "/api" },
      { text: "Repl", link: "/repl" },
      { text: "Examples", link: "https://github.com/windijs/windijs/tree/main/examples" },
    ],
    sidebar: [
      {
        text: "Guide",
        collapsible: false,
        items: [
          {
            text: "What is Windi JS?",
            link: "/guide/what-is-windijs",
          },
          {
            text: "Getting Started",
            link: "/guide/getting-started",
          },
          {
            text: "Configuration",
            link: "/guide/configuration",
          },
          {
            text: "Optimization",
            link: "/guide/optimization",
          },
        ],
      },
      {
        text: "Customization",
        collapsible: false,
        items: [
          {
            text: "Theme",
            link: "/customization/theme",
          },
          {
            text: "Utility",
            link: "/customization/utility",
          },
          {
            text: "Variant",
            link: "/customization/variant",
          },
          {
            text: "Handler",
            link: "/customization/handler",
          },
          {
            text: "Loader",
            link: "/customization/loader",
          },
          {
            text: "Namer",
            link: "/customization/namer",
          },
          {
            text: "Plugins",
            link: "/customization/plugins",
          },
        ],
      },
      {
        text: "Integrations",
        collapsible: false,
        items: [
          {
            text: "API",
            link: "/integrations/api",
          },
          // {
          //   text: "CLI",
          //   link: "/integrations/cli",
          // },
          {
            text: "Vite",
            link: "/integrations/vite",
          },
          // {
          //   text: "Rollup",
          //   link: "/integrations/rollup",
          // },
          // {
          //   text: "Webpack",
          //   link: "/integrations/webpack",
          // },
        ],
      },
      {
        text: "Posts",
        collapsible: false,
        items: [
          {
            text: "How Windi JS Works",
            link: "/posts/how-windijs-works",
          },
          {
            text: "Ways of Using Windi JS",
            link: "/posts/ways-of-using-windijs",
          },
        ],
      },
    ],
    editLink: {
      pattern: "https://github.com/windijs/windijs/edit/main/docs/:path",
      text: "Edit this page on GitHub",
    },
    socialLinks: [
      { icon: "discord", link: "https://discord.gg/nJux6Gb52m" },
      { icon: "github", link: "https://github.com/windijs/windijs" },
    ],
    footer: {
      message: "Released under the MIT License.",
      copyright: "Copyright Â© 2022-present Raven Satir",
    },
  },
});
