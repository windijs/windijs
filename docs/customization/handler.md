---
outline: deep
---

<script setup lang="ts">
import Title from "@/Title.vue"
import { ref } from "vue"

const ts = ref(false)
</script>

<Title head="Handler" v-model:ts="ts" />

The handler refers to the function unit that **handles different props or cases**. We can create any utility by combining different handlers, just like building blocks.

**Note**: Before going deep into these handlers, you should read the [Utility Customization Document](/customization/utility) first, make sure that you understand the structure of the API before continuing.

## Built-in Handlers

### configHandler

### colorHandler

### cssHandler

### numberHandler

### pxHandler

### remHandler

### degHandler

### msHandler

### spacingHandler

### fractionHandler

### callHandler

### setupHandler

### genericHandler

## Handler Helpers

### css

### prop

### meld

### guard

### handler

### isHandler

## Custom Handler

You are able to customize handler by extending built-in handler or recreating a new handler.

### Extending built-in handler

You can extract the handler that you feel redundant into another independent handler. For example:

```js
configHandler(colors, "backgroundColor", "--w-bg-opacity")
```

can be extracted into

```js
export function backgroundColorHandler () {
  return configHandler(colors, "backgroundColor", "--w-bg-opacity")
}
```

Another example, `configHandler` doesn't support handling Array by default, `fontFamilyHandler` convert Array to string first, then pass it to `configHandler`.

<template v-if="ts">

```ts
export function fontFamilyHandler<T extends object> (fonts: T) {
  const cssFonts = {} as {[key in keyof typeof fonts]: string};
  for (const [key, value] of Object.entries(fonts)) {
    cssFonts[key as keyof typeof fonts] = Array.isArray(value) ? value.join(",") : value as string;
  }
  return configHandler(cssFonts as T, "fontFamily");
}
```

</template>

<template v-else>

```js
export function fontFamilyHandler (fonts) {
  const cssFonts = {};
  for (const [key, value] of Object.entries(fonts)) {
    cssFonts[key] = Array.isArray(value) ? value.join(",") : value;
  }
  return configHandler(cssFonts, "fontFamily");
}
```

</template>

One more example, when you use `genericHandler`, you usually want to extract it separately.
This example supports `bg[0x1c1c1e], bg[0xff0]`, ....

<template v-if="ts">

```ts
export function backgroundGenericHandler () {
  return genericHandler<{ [key: string]: StyleObject }>("backgroundColor", (prop) => {
    if (isNumber(prop)) {
      return "#" + (+prop).toString(16);
    }
    return prop;
  });
}
```

</template>

<template v-else>

```js
export function backgroundGenericHandler () {
  return genericHandler("backgroundColor", (prop) => {
    if (isNumber(prop)) {
      return "#" + (+prop).toString(16);
    }
    return prop;
  });
}
```

</template>

### Creating new handler

Basically, the built-in handlers is capable for most cases, but sometimes you may want to create your own handler. In this case, you should use [handler](#handler) api.

A handler in fact is a simple object that has 3 elements, `type`, `meta`, `get`. 

- type
    - Explain: The type of the handler.
    - Type: `"number" | "color" | "config" | "css" | "style" | "call" | "spacing" | "fraction" | "generic" | "setup" | "guard" | "meld" | String`
- meta: 
    - Explain: The meta data of the handler.
    - Type: `object | undefined`
- get:
    - Explain: The function that handles props.
    - Type: `(prop: string) => T`

Let's take [fractionHandler](#fractionhandler) as an example.

```js
import { buildStatic, handler, isFraction, fracToPercent } from "windijs" 

export function fractionHandler (propertyOrBuildFunc) {
  // the build func, return StyleObject
  const build = typeof propertyOrBuildFunc === "function" ? propertyOrBuildFunc : value => buildStatic(propertyOrBuildFunc, value);
  return handler("fraction", p => isFraction(p) ? build(fracToPercent(p)) : undefined)
}
```

And if you are using TypeScript, you can also use type assertion.

```ts
import { buildStatic, handler, isFraction, fracToPercent } from "windijs" 
import type { StyleObject, StyleProperties, BuildFunc, Handler } from "windijs"

export function fractionHandler<T extends object = { [key:string]: StyleObject }> (propertyOrBuildFunc: StyleProperties | StyleProperties[] | BuildFunc) {
  // the build func, return StyleObject
  const build: BuildFunc = typeof propertyOrBuildFunc === "function" ? propertyOrBuildFunc : value => buildStatic(propertyOrBuildFunc, value);
  return {
    type: "fraction",
    get: p => isFraction(p) ? build(fracToPercent(p)) : undefined,
  } as Handler<T>;
}
```