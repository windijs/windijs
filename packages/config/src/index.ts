export * from "./bg";
export * from "./svg";
export * from "./flex";
export * from "./grid";
export * from "./border";
export * from "./effects";
export * from "./behaviors";
export * from "./filter";
export * from "./opacity";
export * from "./sizing";
export * from "./spacing";
export * from "./animation";
export * from "./transform";
export * from "./positioning";
export * from "./interactivity";
export * from "./transitions";
export * from "./typography";
export * from "./keyframes";
export * from "./interface";
