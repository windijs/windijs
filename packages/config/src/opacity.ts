import { range, scales } from "@windijs/shared";

export const opacityConfig = scales(range(0, 21).map(i => i * 5) as [0, 5, 10, 15, 20, 25, 30, 35, 40, 45, 50, 55, 60, 65, 70, 75, 80, 85, 90, 95, 100]);
