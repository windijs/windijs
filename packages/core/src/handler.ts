import { CSSMap, CSSObject, StyleObject, prop } from "@windi/helpers";
import { buildFontSize, buildKeyframes, buildProperty } from "./builder";
import { configHandler, cssHandler, genericHandler, guard } from "./api";
import { prefixAnimation, prefixNotHidden } from "./prefixer";

import { isNumber } from "@windi/shared";

export const animateHandler = <T extends string> (name: T, value: string | CSSObject, keyframes?: Record<string, CSSObject>) => {
  const map: CSSMap = new Map();
  const setValue = (o: CSSObject) => Object.entries(o).map(([k, v]) => map.set(k, v));
  if (keyframes) setValue(buildKeyframes(name, keyframes));
  setValue(typeof value === "string" ? buildProperty([prop`-webkit-animation`, "animation"], value) : prefixAnimation(value));

  return guard(name, cssHandler(map));
};

export function backgroundGenericHandler () {
  return genericHandler<{ [key: string]: StyleObject } & { $: {[key: string]: StyleObject} }>("backgroundColor", (prop) => {
    if (isNumber(prop)) {
      return "#" + (+prop).toString(16);
    }
    return prop;
  });
}

export function divideXReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-divide-x-reverse": "1",
  }));
}

export function divideYReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-divide-y-reverse": "1",
  }));
}

export function fontFamilyHandler<T extends object> (fonts: T) {
  const cssFonts = {} as {[key in keyof typeof fonts]: string};
  for (const [key, value] of Object.entries(fonts)) {
    cssFonts[key as keyof typeof fonts] = Array.isArray(value) ? value.join(",") : value as string;
  }
  return configHandler(cssFonts as T, "fontFamily");
}

export function fontSizeHandler<T extends object> (sizes: T) {
  return configHandler(sizes, value => {
    if (typeof value === "string") return buildFontSize(value);
    if (Array.isArray(value) && value[0]) {
      if (value[1] == null || typeof value[1] === "string") return buildFontSize(value[0], value[1]);
      if (typeof value[1] === "object" && value[1] != null) return buildFontSize(value[0], undefined, value[1]);
    }
  });
}

export function spaceBetweenXReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-space-x-reverse": "1",
  }));
}

export function spaceBetweenYReverseHandler () {
  return cssHandler(prefixNotHidden({
    "--w-space-y-reverse": "1",
  }));
}
