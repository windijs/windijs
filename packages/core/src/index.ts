export * from "./base";
export * from "./api";
export * from "./handler";
export * from "./prefixer";
export * from "./builder";
