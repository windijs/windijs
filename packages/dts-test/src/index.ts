import { defineConfig } from "@windijs/config";

defineConfig({
  utilities: {

  },
});
