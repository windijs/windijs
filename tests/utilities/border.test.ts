import { colors, createUtility } from "../../src";
import { borderRadiusConfig, borderStyleConfig, borderWidthConfig } from "../../src/config/border";
import { opacityConfig } from "../../src/config/opacity";
import { borderColor, borderOpacity, borderRadius, borderStyle, borderWidth } from "../../src/utilities/border";
import { useStaticHandler } from "../../src/utilities/handler";

test("Border Radius", () => {
  const rounded = createUtility("rounded")
    .use(borderRadius(borderRadiusConfig))
    .init();

  // @ts-ignore, hidden property
  expect(rounded.css).toMatchSnapshot("default");
  expect(rounded.full.css).toMatchSnapshot("full");
  expect(rounded[1].css).toMatchSnapshot("1");
  expect(rounded.none.css).toMatchSnapshot("none");
  expect(rounded.sm.css).toMatchSnapshot("sm");
  expect(rounded.md.css).toMatchSnapshot("md");
  expect(rounded.lg.css).toMatchSnapshot("lg");
  expect(rounded.xl.css).toMatchSnapshot("xl");
  expect(rounded.xxl.css).toMatchSnapshot("xxl");
  expect(rounded.xxxl.css).toMatchSnapshot("xxxl");
});

test("Border Style", () => {
  const border = createUtility("border")
    .use(borderStyle(borderStyleConfig))
    .init();
  expect(border.solid.css).toMatchSnapshot("solid");
  expect(border.dashed.css).toMatchSnapshot("dashed");
  expect(border.dotted.css).toMatchSnapshot("dotted");
  expect(border.double.css).toMatchSnapshot("double");
  expect(border.none.css).toMatchSnapshot("none");
});

test("Border With Nested Config", () => {
  const borderStyle = useStaticHandler((handle, styles) => handle(styles, "borderStyle", undefined, true));
  const border = createUtility("border")
    .use(borderStyle({
      solid: "solid",
      nested: {
        DEFAULT: "nested",
        dashed: "dashed",
        nested: {
          double: "double",
        },
      },
    }))
    .init();
  // @ts-ignore
  expect(border.nested.css).toMatchSnapshot();
  expect(border.solid.css).toMatchSnapshot();
  expect(border.nested.dashed.css).toMatchSnapshot();
  expect(border.nested.nested.double.css).toMatchSnapshot();
});

test("Border Opacity", () => {
  const border = createUtility("border")
    .use(borderOpacity(opacityConfig))
    .init();
  expect(border.opacity[0].css).toMatchSnapshot();
  expect(border.opacity[50].css).toMatchSnapshot();
  expect(border.opacity[30].css).toMatchSnapshot();
  expect(border.opacity[100].css).toMatchSnapshot();
});

test("Border Opacity With Different Trigger", () => {
  const border = createUtility("border")
    .use(borderOpacity(opacityConfig, "op"))
    .init();

  expect(border.op[0].css).toMatchSnapshot();
  expect(border.op[50].css).toMatchSnapshot();
  expect(border.op[30].css).toMatchSnapshot();
  expect(border.op[100].css).toMatchSnapshot();
});

test("Border Width", () => {
  const border = createUtility("border")
    .use(borderWidth(borderWidthConfig))
    .init();
  expect(border[0].css).toMatchSnapshot();
  expect(border[4].css).toMatchSnapshot();
  // @ts-ignore
  expect(border.css).toMatchSnapshot();
});

test("Border Color", () => {
  const border = createUtility("border")
    .use(borderColor(colors))
    .init();

  expect(border.current.css).toMatchSnapshot();
  expect(border.white.css).toMatchSnapshot();
  expect(border.blue[700].css).toMatchSnapshot();
  expect(border.red[500].css).toMatchSnapshot();
});

test("Border Meta", () => {
  const rounded = createUtility("rounded")
    .use(borderRadius(borderRadiusConfig))
    .init();

  const border = createUtility("border")
    .use(borderColor(colors))
    .use(borderWidth(borderWidthConfig))
    .use(borderOpacity(opacityConfig, "op"))
    .use(borderStyle(borderStyleConfig))
    .init();

  // @ts-ignore, hidden property
  expect(rounded.meta).toMatchSnapshot();
  expect(rounded.sm.meta).toMatchSnapshot();
  // @ts-ignore, hidden property
  expect(border.meta).toMatchSnapshot();
  expect(border[2].meta).toMatchSnapshot();
  expect(border.black.meta).toMatchSnapshot();
  expect(border.blue[400].meta).toMatchSnapshot();
  expect(border.dashed.meta).toMatchSnapshot();
  expect(border.op[50].meta).toMatchSnapshot();
});
