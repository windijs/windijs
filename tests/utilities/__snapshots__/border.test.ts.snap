// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Border Color 1`] = `
Object {
  "borderColor": "currentColor",
}
`;

exports[`Border Color 2`] = `
Object {
  "--w-border-opacity": "1",
  "borderColor": "rgba(255, 255, 255, var(--w-border-opacity))",
}
`;

exports[`Border Color 3`] = `
Object {
  "--w-border-opacity": "1",
  "borderColor": "rgba(29, 78, 216, var(--w-border-opacity))",
}
`;

exports[`Border Color 4`] = `
Object {
  "--w-border-opacity": "1",
  "borderColor": "rgba(239, 68, 68, var(--w-border-opacity))",
}
`;

exports[`Border Opacity 1`] = `
Object {
  "--w-border-opacity": "0",
}
`;

exports[`Border Opacity 2`] = `
Object {
  "--w-border-opacity": "0.5",
}
`;

exports[`Border Opacity 3`] = `
Object {
  "--w-border-opacity": "0.3",
}
`;

exports[`Border Opacity 4`] = `
Object {
  "--w-border-opacity": "1",
}
`;

exports[`Border Opacity With Different Trigger 1`] = `
Object {
  "--w-border-opacity": "0",
}
`;

exports[`Border Opacity With Different Trigger 2`] = `
Object {
  "--w-border-opacity": "0.5",
}
`;

exports[`Border Opacity With Different Trigger 3`] = `
Object {
  "--w-border-opacity": "0.3",
}
`;

exports[`Border Opacity With Different Trigger 4`] = `
Object {
  "--w-border-opacity": "1",
}
`;

exports[`Border Opacity With Proxy Config 1`] = `
Object {
  "--w-border-opacity": "0",
}
`;

exports[`Border Opacity With Proxy Config 2`] = `
Object {
  "--w-border-opacity": "0.5",
}
`;

exports[`Border Opacity With Proxy Config 3`] = `
Object {
  "--w-border-opacity": "0.3",
}
`;

exports[`Border Opacity With Proxy Config 4`] = `
Object {
  "--w-border-opacity": "1",
}
`;

exports[`Border Opacity With Proxy Config 5`] = `
Object {
  "--w-border-opacity": "0.12",
}
`;

exports[`Border Opacity With Proxy Config 6`] = `
Object {
  "--w-border-opacity": "0.7",
}
`;

exports[`Border Radius: 1 1`] = `
Object {
  "borderRadius": "100%",
}
`;

exports[`Border Radius: default 1`] = `
Object {
  "borderRadius": "0.25rem",
}
`;

exports[`Border Radius: full 1`] = `
Object {
  "borderRadius": "9999px",
}
`;

exports[`Border Radius: lg 1`] = `
Object {
  "borderRadius": "0.5rem",
}
`;

exports[`Border Radius: md 1`] = `
Object {
  "borderRadius": "0.375rem",
}
`;

exports[`Border Radius: none 1`] = `
Object {
  "borderRadius": "0px",
}
`;

exports[`Border Radius: sm 1`] = `
Object {
  "borderRadius": "0.125rem",
}
`;

exports[`Border Radius: xl 1`] = `
Object {
  "borderRadius": "0.75rem",
}
`;

exports[`Border Radius: xxl 1`] = `
Object {
  "borderRadius": "1rem",
}
`;

exports[`Border Radius: xxxl 1`] = `
Object {
  "borderRadius": "1.5rem",
}
`;

exports[`Border Style: dashed 1`] = `
Object {
  "borderStyle": "dashed",
}
`;

exports[`Border Style: dotted 1`] = `
Object {
  "borderStyle": "dotted",
}
`;

exports[`Border Style: double 1`] = `
Object {
  "borderStyle": "double",
}
`;

exports[`Border Style: none 1`] = `
Object {
  "borderStyle": "none",
}
`;

exports[`Border Style: solid 1`] = `
Object {
  "borderStyle": "solid",
}
`;

exports[`Border Width 1`] = `
Object {
  "borderWidth": "0px",
}
`;

exports[`Border Width 2`] = `
Object {
  "borderWidth": "4px",
}
`;

exports[`Border Width 3`] = `
Object {
  "borderWidth": "1px",
}
`;

exports[`Border With Nested Config 1`] = `
Object {
  "borderStyle": "nested",
}
`;

exports[`Border With Nested Config 2`] = `
Object {
  "borderStyle": "solid",
}
`;

exports[`Border With Nested Config 3`] = `
Object {
  "borderStyle": "dashed",
}
`;

exports[`Border With Nested Config 4`] = `
Object {
  "borderStyle": "double",
}
`;
